ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/ExpressionMetadata.java

```diff
@@ -62,16 +64,35 @@ public Expression getExpression() {
     return expression;
   }
 
-  public Object evaluate(final GenericRow row) {
+  /**
+   * Evaluate the expression against the supplied {@code row}.
+   *
+   * <p>On error the supplied {@code logger} is called with the details of the error and the method
+   * return {@code null}.
+   *
+   * @param row the row of data to evaluate the expression against.
+   * @param defaultValue the value to return if an exception is thrown.
+   * @param logger an optional logger to log errors to. If not supplied the method throws on error.
+   * @param errorMsg called to get the text for the logged error.
+   * @return the result of the evaluation.
+   */
+  public Object evaluate(
+      final GenericRow row,
+      final Object defaultValue,
+      final ProcessingLogger logger,
+      final Supplier<String> errorMsg
+  ) {
     try {
       return expressionEvaluator.evaluate(getParameters(row));
     } catch (final InvocationTargetException e) {
-      throw new KsqlException(e.getCause().getMessage(), e.getCause());
+      logger.error(RecordProcessingError.recordProcessingError(errorMsg.get(), e.getCause(), row));
```

Comment:
I think there's an existing configuration about whether to hard-fail on error or log-and-continue. If there is, we should respect that, and if there isn't we should add one so people can control if they want their application skipping events

Reply:
It's an interesting point... 

However, the changes I've made in this PR close a few holes where we invoke `evaluate` but don't protect it within a try/catch(log) block. 

The existing places that do this try/catch(log) block do NOT honour any streams exception handling config on whether to exist or log-and-continue.

Hence, while I think it's an interesting point, I also think its outside the scope of this change.  A log of our code / UDFs etc throw NPEs and a variety of other exceptions. We'd need to clean all those up before having a query terminate due to bad data...

Reply:
Talked offline - @agavra's happy with this.